### PR TITLE
Build development version

### PR DESCRIPTION
**Pull request by ocefpaf**
@mrocklin can you take a quick look?
I am using the tarball archive and not the actual branch, is that OK?

After merging this should be available for download with `conda install -c conda-forge/dev geopandas`

Closes #30

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,21 +1,22 @@
-{% set version = "0.3.0" %}
+{% set version = "1.0.0.dev" %}
 
 package:
   name: geopandas
   version: {{ version }}
 
 source:
-  url: https://github.com/geopandas/geopandas/releases/download/v{{ version }}/geopandas-{{ version }}.tar.gz
-  sha256: e63bb32a3e516d8c9bcd149c22335575defdc5896c8bdf15c836608f152a920b
```
Comment (mrocklin) [inline recipe/meta.yaml]: How do we update builds in either of the following cases:

1.  geopandas master (without cython) wants to release?  

    We revert these changes?
2.  geopandas geopandas-cython wants to push a new dev build?
    
    We just update the sha here?  Perhaps also bump the dev build number?

**Comment by mrocklin** [comment]
cc @jorisvandenbossche

File: recipe/meta.yaml [hunk 2]
```diff
@@ -1,21 +1,22 @@
-{% set version = "0.3.0" %}
+{% set version = "1.0.0.dev" %}
```
Comment (mrocklin) [inline recipe/meta.yaml]: Should this version be less?  Somewhere near 0.4.0?

File: recipe/meta.yaml (hunk 1 above)
Comment (ocefpaf) [reply recipe/meta.yaml]: > geopandas master (without cython) wants to release?

No need to revert but substitute this with the GitHub URL, but if the branch is more up-to-date than this archived release I can put it in there now.

> geopandas geopandas-cython wants to push a new dev build?

I would recommend bumping the build number or, if geopandas uses versioneer, just use that version as a new release.

File: recipe/meta.yaml (hunk 2 above)
Comment (ocefpaf) [reply recipe/meta.yaml]: I just got what is in the archived tarball BTW.

**Comment by jorisvandenbossche** [comment]
@ocefpaf Thanks a lot for this! 
Shall we target this PR to a 'dev' branch, so if we still do a 'normal' geopandas release we don't have to undo the changes?

**Comment by ocefpaf** [comment]
> Shall we target this PR to a 'dev' branch, so if we still do a 'normal' geopandas release we don't have to undo the changes?

I created the `dev` branch and did not target it. My bad :flushed:

File: recipe/meta.yaml (hunk 2 above)
Comment (jorisvandenbossche) [reply recipe/meta.yaml]: I don't think this will work with setting at 1.0.0.dev, as that tag is from some time ago. I think we always want to build the current latest version.

If that is not possible, and for conda-forge there always needs to be a tag to build, we can add some new tags like alpha 1, 2, 3, ... etc each time we want to update this build.

**Comment by jorisvandenbossche** [comment]
You can change that in a PR, no need to close

File: recipe/meta.yaml (hunk 1 above)
Comment (jorisvandenbossche) [reply recipe/meta.yaml]: Yes, geopandas uses versioneer, so the last 'version' is 

```
In [2]: geopandas.__version__
Out[2]: '1.0.0.dev+100.g170c235'
```

File: recipe/meta.yaml (hunk 2 above)
Comment (ocefpaf) [reply recipe/meta.yaml]: [`rasterio`](https://github.com/conda-forge/rasterio-feedstock/blob/dev/recipe/meta.yaml#L1) and [`fiona`](https://github.com/conda-forge/fiona-feedstock/pull/59/files) as using alpha releases so, b/c `geopandas` users are familiar with those libraries, we should do the same here instead of using the version + git hash from versioneer. What do you think?

Comment (jorisvandenbossche) [reply recipe/meta.yaml]: Yes, although they use it for their 'main' release as well, while I think for now we don't want people to automatically install this dev version. 
It's a bit more work to tag each time to update this build, but also not that much work. So either is fine with me.

File: recipe/meta.yaml [hunk 3]
```diff
@@ -1,21 +1,24 @@
...
 
 package:
   name: geopandas
   version: {{ version }}
 
 source:
-  url: https://github.com/geopandas/geopandas/releases/download/v{{ version }}/geopandas-{{ version }}.tar.gz
-  sha256: e63bb32a3e516d8c9bcd149c22335575defdc5896c8bdf15c836608f152a920b
+  git_url: https://github.com/geopandas/geopandas.git
+  git_branch: geopandas-cython
```
Comment (ocefpaf) [inline recipe/meta.yaml]: @jorisvandenbossche and @mrocklin how about we do this in a simple way then. By using the branch directly all you need to do to create a new package is to bump the build number. The first is up to you, `1.0.0.dev` or `0.4.0.dev`?

Comment (jorisvandenbossche) [reply recipe/meta.yaml]: 1.0.0.dev is on the geopandas-cython branch (0.4.0.dev does not exist)

**Comment by ocefpaf** [comment]
Looks like Windows needs a fallback option to find geos.

**Comment by jorisvandenbossche** [comment]
Regarding windows, I think I experienced the same when trying to get it working on appveyor for tests: https://github.com/geopandas/geopandas/pull/570 (so but not sure how to solve it)

**Comment by ocefpaf** [comment]
> Regarding windows, I think I experienced the same when trying to get it working on appveyor for tests: https://github.com/geopandas/geopandas/pull/570 (so but not sure how to solve it)

One way to do that is to let users specify it. I guess that a pattenr like [this one](https://github.com/Toblerity/Fiona/blob/master/setup.py#L106-L126) or environmental variables would do. See my attempt in https://github.com/conda-forge/geopandas-feedstock/pull/30/commits/5c2e8e43b44c037e6ff860c08ff84205e35be5bc, if you are OK with that I can send a PR upstream.

File: recipe/stdbool.h [hunk 4]
```diff
@@ -0,0 +1,3 @@
+typedef int bool;
+#define false 0
+#define true 1
```
Comment (ocefpaf) [inline recipe/stdbool.h]: @jorisvandenbossche and @jorisvandenbossche the compilation on Windows and vc 9 requires msinttypes for `stdint.h` and `stdbool.h`. However, even with those two the compilation fails further down the road and I am not sure if it is worth patching the code or just abandoning vc 9 (that means abandoning Python 2.7).

File: recipe/meta.yaml [hunk 5]
```diff
@@ -7,20 +7,29 @@ package:
...
   git_url: https://github.com/geopandas/geopandas.git
   git_branch: geopandas-cython
+  patches:
+    - geos_header.patch
 
 build:
   number: 0
-  script: python setup.py install --single-version-externally-managed --record record.txt --with-cython
+  script:
+    - python setup.py install --single-version-externally-managed --record record.txt --with-cython  # [not win]
```
Comment (jorisvandenbossche) [inline recipe/meta.yaml]: What is the `[not win]` on this line for?

Comment (ocefpaf) [reply recipe/meta.yaml]: B/c I added a `bld.bat` with all the extra stuff that only Windows needs.

File: recipe/stdbool.h (hunk 4 above)
Comment (jorisvandenbossche) [reply recipe/stdbool.h]: I would find dropping 2.7 support a bit unfortunate, but probably OK to leave it for now to get this moving. 
It also seems the failure is only in the STRTree related code, which is not essential for basic usage (but of course you quickly want to do a spatial join)

**Comment by jorisvandenbossche** [comment]
Yes, such a patch to find the geos library would be welcome to send upstream

**Comment by jorisvandenbossche** [comment]
I merged the PR on geopandas, so the patch here can be removed?

**Comment by ocefpaf** [comment]
> I merged the PR on geopandas, so the patch here can be removed?

Done. I guess this is OK to be merged now.

**Comment by jorisvandenbossche** [comment]
Great! Thanks a lot, this will make it much easier to test this out!

**Comment by ocefpaf** [comment]
Cool. They way it is set all you need to do is to bump the build number to get a new build with the latest checkout of the `geopandas-cython` branch. Note that to install this one needs to type:

```
conda install -c conda-forge/dev geopandas
```

**Comment by jorisvandenbossche** [comment]
Small correction, it is apparently conda-forge/label/dev. \*

And that works!

```
joris@joris-XPS-13-9350:~/scipy$ conda create -n geopandas-test python=3 geopandas -c conda-forge
...
joris@joris-XPS-13-9350:~/scipy$ act geopandas-test
(geopandas-test) joris@joris-XPS-13-9350:~/scipy$ conda install -c conda-forge/label/dev geopandas
Fetching package metadata .................
Solving package specifications: .

Package plan for installation in environment /home/joris/miniconda3/envs/geopandas-test:

The following packages will be UPDATED:

    geopandas: 0.3.0-py36_0 conda-forge --> 1.0.0.dev-py36_0 conda-forge/label/dev

Proceed ([y]/n)? 

geopandas-1.0. 100% |################################################################################################################################################################################################################################| Time: 0:00:15 103.44 kB/s

(geopandas-test) joris@joris-XPS-13-9350:~/scipy$ python
Python 3.6.3 | packaged by conda-forge | (default, Nov  4 2017, 10:10:56) 
[GCC 4.8.2 20140120 (Red Hat 4.8.2-15)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import geopandas
>>> geopandas.__version__
'1.0.0.dev+101.g1c62492'
>>> df = geopandas.read_file(geopandas.datasets.get_path("naturalearth_cities"))
>>> df.head()
I am densified (5 elements)
           name                                     geometry
0  Vatican City  POINT (12.45338654497177 41.90328217996012)
1    San Marino    POINT (12.44177015780014 43.936095834768)
2         Vaduz  POINT (9.516669472907267 47.13372377429357)
3    Luxembourg  POINT (6.130002806227083 49.61166037912108)
4       Palikir  POINT (158.1499743237623 6.916643696007725)
>>> df.geometry._geometry_array
<geopandas.vectorized.GeometryArray object at 0x7fa3fd636940>
```


\* using the wrong channel label is a bit confusing, as it doesn't give you an error that it doesn't exist, but just a strange warning and then continues while ignoring the channel:

```
(geopandas-test) joris@joris-XPS-13-9350:~/scipy$ conda install geopandas -c conda-forge/dev
Fetching package metadata ....
WARNING: The remote server could not find the noarch directory for the
requested channel with url: https://conda.anaconda.org/conda-forge/dev

It is possible you have given conda an invalid channel. Please double-check
your conda configuration using `conda config --show`.

If the requested url is in fact a valid conda channel, please request that the
channel administrator create `noarch/repodata.json` and associated
`noarch/repodata.json.bz2` files, even if `noarch/repodata.json` is empty.
$ mkdir noarch
$ echo '{}' > noarch/repodata.json
$ bzip2 -k noarch/repodata.json
.............
Solving package specifications: .

# All requested packages already installed.
# packages in environment at /home/joris/miniconda3/envs/geopandas-test:
#
geopandas                 0.3.0                    py36_0    conda-forge
```

**Comment by ocefpaf** [comment]
Yep. I forgot about the label, just fixed it above. Thanks!

PS: I agree that the wrong channel should result in an error.
